core/solver/chebyshev.cpp

```diff
@@ -0,0 +1,341 @@
+// SPDX-FileCopyrightText: 2025 The Ginkgo authors
+//
+// SPDX-License-Identifier: BSD-3-Clause
+
+#include "ginkgo/core/solver/chebyshev.hpp"
+
+#include <ginkgo/core/base/precision_dispatch.hpp>
+#include <ginkgo/core/matrix/dense.hpp>
+#include <ginkgo/core/solver/solver_base.hpp>
+
+#include "core/config/solver_config.hpp"
+#include "core/distributed/helpers.hpp"
+#include "core/solver/chebyshev_kernels.hpp"
+#include "core/solver/ir_kernels.hpp"
+#include "core/solver/solver_base.hpp"
+#include "core/solver/solver_boilerplate.hpp"
+#include "core/solver/update_residual.hpp"
+
+
+namespace gko {
+namespace solver {
+namespace chebyshev {
+namespace {
+
+
+GKO_REGISTER_OPERATION(initialize, ir::initialize);
+GKO_REGISTER_OPERATION(init_update, chebyshev::init_update);
+GKO_REGISTER_OPERATION(update, chebyshev::update);
+
+
+}  // anonymous namespace
+}  // namespace chebyshev
+
+template <typename ValueType>
+typename Chebyshev<ValueType>::parameters_type Chebyshev<ValueType>::parse(
+    const config::pnode& config, const config::registry& context,
+    const config::type_descriptor& td_for_child)
+{
+    auto params = solver::Chebyshev<ValueType>::build();
+    common_solver_parse(params, config, context, td_for_child);
+    if (auto& obj = config.get("foci")) {
+        auto arr = obj.get_array();
+        if (arr.size() != 2) {
+            GKO_INVALID_CONFIG_VALUE("foci", "must contain two elements");
+        }
+        params.with_foci(gko::config::get_value<ValueType>(arr.at(0)),
+                         gko::config::get_value<ValueType>(arr.at(1)));
+    }
+    if (auto& obj = config.get("default_initial_guess")) {
+        params.with_default_initial_guess(
+            gko::config::get_value<solver::initial_guess_mode>(obj));
+    }
+    return params;
+}
+
+
+template <typename ValueType>
+Chebyshev<ValueType>::Chebyshev(const Factory* factory,
+                                std::shared_ptr<const LinOp> system_matrix)
+    : EnableLinOp<Chebyshev>(factory->get_executor(),
+                             gko::transpose(system_matrix->get_size())),
+      EnablePreconditionedIterativeSolver<ValueType, Chebyshev<ValueType>>{
+          std::move(system_matrix), factory->get_parameters()},
+      parameters_{factory->get_parameters()}
+{
+    this->set_default_initial_guess(parameters_.default_initial_guess);
+    center_ = (std::get<0>(parameters_.foci) + std::get<1>(parameters_.foci)) /
```

Comment:
maybe we should have some checks on the `foci` parameter, i.e. `foci[0,1] != 0, foci[0] <= foci[1]`.

Reply:
> In L1 scalar Jacobi on SPD matrix, we use 0, 1 here, which provided by the math property

Can you explain this a bit more?

Reply:
Ok, I see, you meant the range of the eigenvalues under preconditioning. But I don't understand what this has to do with checking the input values for the foci parameter.

Reply:
It only checks the real(foci[1]) >= real(foci[0]) now

Reply:
I would think that's enough.